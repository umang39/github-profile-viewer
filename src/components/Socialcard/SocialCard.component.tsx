import React, { ReactElement } from "react";
import { Avatar } from "@material-ui/core";
import { Grid, Link, Typography } from "@mui/material";
import { Box } from "@mui/system";
import { socialCardInterface, useSocialStyle } from "@components/index";
import MailIcon from "@mui/icons-material/Mail";
import GitHubIcon from "@mui/icons-material/GitHub";
import { theme } from "@src/theme";
import { ButtonComponent, EmptySection } from "@components/index";

const SocialCard = (props: socialCardInterface): ReactElement => {
  const { user, followersHandler, followHandler, isFollow } = props;
  const {
    avatar_url,
    login,
    location,
    followers,
    following,
    bio,
    email,
    html_url,
    blog,
  } = user;
  const socialStyles = useSocialStyle();
  return (
    <>
      <Grid
        container
        style={{ width: "70%" }}
        className={socialStyles.container}
        boxShadow={10}
      >
        <Grid item md={6} sm={12}>
          <Grid container justifyContent="center">
            <Grid item sm={12}>
              <Grid container justifyContent="center">
                <Grid item xs={12} className={socialStyles.avatarWrapper}>
                  <Avatar
                    alt="Cindy Baker"
                    src={avatar_url}
                    className={socialStyles.avatar}
                  />
                </Grid>
                <Grid item xs={12} className={socialStyles.follow}>
                  <Typography
                    marginTop={1}
                    fontSize={theme.spacing(3.75)}
                    className={(socialStyles.head, "line-break")}
                    fontWeight={550}
                  >
                    {login}
                  </Typography>
                </Grid>

                <Grid item xs={12} className={socialStyles.follow}>
                  {location && <Typography variant="h5">{location}</Typography>}
                </Grid>
                {localStorage.getItem("username") !== login && !isFollow ? (
                  <Grid
                    item
                    xs={12}
                    className={socialStyles.followBtn}
                    marginY={1}
                  >
                    <ButtonComponent
                      type="button"
                      value="Follow"
                      id={login}
                      btnHandler={followHandler}
                      isDisabled={false}
                    />
                  </Grid>
                ) : (
                  <></>
                )}
                <Grid
                  item
                  sm={6}
                  xs={12}
                  className={socialStyles.followContainer}
                >
                  <Grid container className={socialStyles.followerContainer}>
                    <Grid item xs={12}>
                      <Typography variant="h3" className={socialStyles.follow}>
                        {followers}
                      </Typography>
                    </Grid>
                    <Grid item>
                      <Typography
                        variant="h4"
                        className={(socialStyles.follow, socialStyles.follower)}
                        onClick={(e: React.SyntheticEvent) => {
                          return followersHandler(e, login);
                        }}
                      >
                        Followers
                      </Typography>
                    </Grid>
                  </Grid>
                </Grid>
                <Grid
                  item
                  sm={6}
                  xs={12}
                  className={socialStyles.followContainer}
                >
                  <Grid container>
                    <Grid item xs={12} className={socialStyles.follow}>
                      <Typography variant="h3">{following}</Typography>
                    </Grid>
                    <Grid item xs={12}>
                      <Typography variant="h4" className={socialStyles.follow}>
                        Following
                      </Typography>
                    </Grid>
                  </Grid>
                </Grid>
                <Grid item md={12}>
                  <Grid container marginTop={1}>
                    <Grid item md={3} xs={6}>
                      <Typography>
                        <Link href={html_url} color="inherit">
                          <GitHubIcon
                            id="github"
                            style={{
                              fontSize: theme.spacing(3),
                            }}
                          />
                        </Link>
                      </Typography>
                    </Grid>
                    <Grid md={3} xs={6} item>
                      <Typography>
                        {email && (
                          <Link href={`mailto:${email}`} color="inherit">
                            <MailIcon
                              id="mail"
                              style={{
                                fontSize: theme.spacing(3),
                              }}
                            />
                          </Link>
                        )}
                      </Typography>
                    </Grid>
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </Grid>
        <Grid item md={6} xs={12}>
          <Grid container>
            <Grid item xs={12}>
              <Typography
                fontSize={theme.spacing(3.75)}
                fontWeight={550}
                className={socialStyles.head}
              >
                Bio
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <Box>
                <Typography fontSize={15} className="line-break">
                  {bio ? bio : <EmptySection title="bio" />}
                </Typography>
              </Box>
            </Grid>
            {blog && (
              <>
                <Grid item xs={12} marginTop={4}>
                  <Typography
                    fontSize={theme.spacing(3.75)}
                    fontWeight={550}
                    className={socialStyles.head}
                  >
                    Blog
                  </Typography>
                </Grid>
                <Grid item xs={12}>
                  <Box>
                    <Typography fontSize={15} className="line-break">
                      {blog}
                    </Typography>
                  </Box>
                </Grid>
              </>
            )}
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

export default SocialCard;
