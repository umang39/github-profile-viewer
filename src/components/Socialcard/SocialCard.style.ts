import { makeStyles } from "@material-ui/core";
import { grey } from "@mui/material/colors";
import { theme } from "../../theme";
export const useSocialStyle = makeStyles({
  container: {
    background: `linear-gradient(120deg ,${grey[300]} 30%,${grey[100]})`,
    backgroundPosition: "right bottom",
    backgroundSize: "150% 150%",
    borderRadius: 20,
    padding: theme.spacing(1),
    color: theme.palette.primary.main,
    height: "max-content",
    [theme.breakpoints.up("md")]: {
      padding: theme.spacing(3),
    },
    [theme.breakpoints.down("md")]: {
      margin: theme.spacing(1),
    },
  },
  followBtn: {
    textAlign: "center",
    [theme.breakpoints.up("md")]: {
      textAlign: "left",
    },
  },
  follower: {
    textAlign: "left",
    borderBottom: `${theme.spacing(0.2)} solid transparent`,
    [theme.breakpoints.down("md")]: {
      textAlign: "center",
    },
    "&:hover": {
      cursor: "pointer",
      borderBottom: `${theme.spacing(0.2)} solid ${theme.palette.primary.main}`,
    },
  },
  follow: {
    textAlign: "center",
    [theme.breakpoints.up("md")]: {
      textAlign: "unset",
    },
  },
  followerContainer: {
    justifyContent: "center",
    [theme.breakpoints.up("md")]: {
      justifyContent: "unset",
    },
  },
  followContainer: {
    margin: "10rem 0",
    [theme.breakpoints.up("xs")]: {
      margin: "unset",
    },
  },
  avatarWrapper: {
    [theme.breakpoints.up("md")]: {
      display: "unset",
      justifyContent: "unset",
    },
    display: "flex",
    justifyContent: "center",
  },
  head: {
    color: "#616161",
    fontWeight: "bolder",
  },
  user: {
    marginTop: 10,
  },
  avatar: {
    height: 100,
    width: 100,
  },
});
