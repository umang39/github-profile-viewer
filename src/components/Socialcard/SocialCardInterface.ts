import React from "react";

export interface socialCardInterface {
  user: {
    login: string;
    followers: number;
    following: number;
    bio: string | null;
    location: string | null;
    email: string | null;
    name: string | null;
    avatar_url: string;
    html_url: string;
    blog: string | null;
  };
  isFollow?: boolean;
  followHandler?: (e: React.SyntheticEvent) => void;
  isFollowingUser?: (user: string) => void;
  followersHandler: (e: React.SyntheticEvent, login: string) => void;
}
