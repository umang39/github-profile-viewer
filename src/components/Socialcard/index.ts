import SocialCard from "@components/Socialcard/SocialCard.component";
import { useSocialStyle } from "@components/Socialcard/SocialCard.style";
import { socialCardInterface } from "@components/Socialcard/SocialCardInterface";

export { SocialCard, useSocialStyle };
export type { socialCardInterface };
