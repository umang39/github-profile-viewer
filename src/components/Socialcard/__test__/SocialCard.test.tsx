import React from "react";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import { SocialCard } from "../..";
import renderer from "react-test-renderer";
import { socialCardInterface } from "@components/Socialcard/index";
afterEach(cleanup);
const authUser = {
  user: {
    avatar_url: "https://avatars.githubusercontent.com/u/92724856?v=4",
    login: "umang39",
    location: null,
    followers: 3,
    following: 0,
    bio: null,
    name: null,
    email: null,
    html_url: "https://github.com/umang39",
    blog: null,
  },
  followersHandler: jest.fn(),
};
const searchedUser = {
  user: {
    avatar_url: "https://avatars.githubusercontent.com/u/92724856?v=4",
    login: "xyz",
    location: "new york",
    followers: 3,
    following: 0,
    bio: null,
    name: null,
    email: "xyz@gmail.com",
    html_url: "https://github.com/umang39",
    blog: null,
  },
  followersHandler: jest.fn(),
};
const socialCardHandler = (props: socialCardInterface = authUser) => {
  return render(
    <SocialCard user={props.user} followersHandler={props.followersHandler} />
  );
};
it("social card  running without crashing", () => {
  socialCardHandler();
  const followers = screen.queryByText(3)?.innerHTML;
  const username = screen.queryByText("umang39")?.innerHTML;
  expect(username).toBe(authUser.user.login);
  expect(followers).toBe(authUser.user.followers.toString());
});

it("social card for searched users running without crashing", () => {
  socialCardHandler({
    user: searchedUser.user,
    followersHandler: searchedUser.followersHandler,
  });
  const username = screen.queryByText("xyz")?.innerHTML;
  expect(username).toBe(searchedUser.user.login);
});
it("mail is successfully sending", () => {
  socialCardHandler({
    user: searchedUser.user,
    followersHandler: searchedUser.followersHandler,
  });
  const email = screen.queryAllByRole("link")[1].getAttribute("href");
  expect(email).toBe(`mailto:${searchedUser.user.email.toString()}`);
});
it("follow button running without crashing", () => {
  const { container } = socialCardHandler({
    user: searchedUser.user,
    followersHandler: searchedUser.followersHandler,
  });
  fireEvent.click(container.querySelector("#xyz") as HTMLInputElement);
  expect(container.querySelector("#xyz")).toBeInTheDocument();
});
it("followers Button clicked succesfully", () => {
  const { container } = socialCardHandler({
    user: searchedUser.user,
    followersHandler: searchedUser.followersHandler,
  });
  fireEvent.click(container.querySelector("h4") as HTMLInputElement);
  expect(searchedUser.followersHandler).toBeCalled();
});
it("matches snapshot", () => {
  const tree = renderer
    .create(
      <SocialCard
        user={authUser.user}
        followersHandler={authUser.followersHandler}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
