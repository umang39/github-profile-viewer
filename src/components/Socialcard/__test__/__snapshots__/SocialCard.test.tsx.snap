// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined makeStyles-container-61 css-ibhn6n-MuiGrid-root"
  style={
    Object {
      "width": "70%",
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6 css-sxdago-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-1lym95h-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 css-16nn1fs-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-1lym95h-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 makeStyles-avatarWrapper-67 css-1em9ut6-MuiGrid-root"
          >
            <div
              className="MuiAvatar-root MuiAvatar-circular makeStyles-avatar-70"
            >
              <img
                alt="Cindy Baker"
                className="MuiAvatar-img"
                src="https://avatars.githubusercontent.com/u/92724856?v=4"
              />
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 makeStyles-follow-64 css-1em9ut6-MuiGrid-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 line-break css-1ix858w-MuiTypography-root"
            >
              umang39
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 makeStyles-follow-64 css-1em9ut6-MuiGrid-root"
          />
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 makeStyles-followBtn-62 css-36dx3p-MuiGrid-root"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btn-71"
              disabled={false}
              id="umang39"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Follow
              </span>
            </button>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 makeStyles-followContainer-66 css-wdtwov-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined makeStyles-followerContainer-65 css-11lq3yg-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
              >
                <h3
                  className="MuiTypography-root MuiTypography-h3 makeStyles-follow-64 css-gepadz-MuiTypography-root"
                >
                  3
                </h3>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <h4
                  className="MuiTypography-root MuiTypography-h4 makeStyles-follower-63 css-5lbw0b-MuiTypography-root"
                  onClick={[Function]}
                >
                  Followers
                </h4>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 makeStyles-followContainer-66 css-wdtwov-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 makeStyles-follow-64 css-1em9ut6-MuiGrid-root"
              >
                <h3
                  className="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
                >
                  0
                </h3>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
              >
                <h4
                  className="MuiTypography-root MuiTypography-h4 makeStyles-follow-64 css-5lbw0b-MuiTypography-root"
                >
                  Following
                </h4>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-12 css-15fi2ac-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-1sthlmf-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-1h9mey6-MuiGrid-root"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  <a
                    className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
                    href="https://github.com/umang39"
                    onBlur={[Function]}
                    onFocus={[Function]}
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="GitHubIcon"
                      focusable="false"
                      id="github"
                      style={
                        Object {
                          "fontSize": "3rem",
                        }
                      }
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
                      />
                    </svg>
                  </a>
                </p>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 css-1h9mey6-MuiGrid-root"
              >
                <p
                  className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1tp3ba8-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 makeStyles-head-68 css-sopyuc-MuiTypography-root"
        >
          Bio
        </p>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
      >
        <div
          className="MuiBox-root css-0"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 line-break css-19t76t3-MuiTypography-root"
          >
            <h2
              className="MuiTypography-root MuiTypography-h2 makeStyles-text-72 css-cq2q8w-MuiTypography-root"
            >
              Oops ! You have no 
              bio
            </h2>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
